src/lib.rs

```diff
@@ -248,31 +258,307 @@ impl<T, const N: usize> ArrayVec<T, { N }> {
+    /// assert_eq!(&vector, [24, 42].as_ref());
+    /// ```
+    ///
+    /// But when the vector's full, we get a different result:
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector = ArrayVec::from([
+    ///     "He".to_owned(),
+    ///     "ya".to_owned(),
+    /// ]);
+    ///
+    /// let out = vector
+    ///     .force_insert(1, "llo".to_owned())
+    ///     .unwrap();
+    ///
+    /// assert_eq!(&out, "ya");
+    /// assert_eq!(&vector, ["He".to_owned(), "llo".to_owned()].as_ref());
+    /// ```
+    pub fn force_insert(&mut self, index: usize, item: T) -> Option<T> {
+        let len = self.len();
+
+        // bound checks
+        if index > len {
+            out_of_bounds!("force_insert", index, len);
+        }
+
+        let result;
+
+        if self.is_full() {
+            // bound checks (here: N == len)
+            if index >= N {
+                out_of_bounds!("force_insert", index, len);
+            }
+
+            unsafe {
+                // Store the last item before it's removed.
+                result = Some(ptr::read(self.as_ptr().add(len - 1)));
+
+                // The last element should be removed so we shouldn't
+                // copy it.
+                self.insert_unchecked_keep_len(index, item, len - 1);
+            }
+        } else {
+            // bound checks
+            if index > len {
+                out_of_bounds!("force_insert", index, len);
+            }
+
+            // Since nothing's going to be removed, the vector's size
+            // is going to be increased and nothing will be returned.
+            unsafe {
+                self.insert_unchecked(index, item, len);
+                result = None;
+            }
+        }
+
+        result
+    }
+
+    /// Insert an item into the vector without checking if the index is
+    /// valid or if the vector isn't full or the vector's length.
+    ///
+    /// # Safety
+    ///
+    /// If you plan on using this function, you need to check for the
+    /// 3 previously mentioned conditions yourself before calling this
+    /// method.
+    #[inline]
+    pub unsafe fn insert_unchecked(
+        &mut self,
+        index: usize,
+        item: T,
+        len: usize,
+    ) {
+        self.insert_unchecked_keep_len(index, item, len);
+        self.set_len(len + 1);
+    }
+
+    /// Insert an item into the vector without checking if the index is
+    /// valid or if the vector isn't full or the vector's length and
+    /// without incrementing the vector's length.
+    ///
+    /// # Safety
+    ///
+    /// If you plan on using this function, you need to check for the
+    /// 3 previously mentioned conditions yourself before calling this
+    /// method. You also need to increment the vector's length afterward
+    /// yourself.
+    pub unsafe fn insert_unchecked_keep_len(
+        &mut self,
+        index: usize,
+        item: T,
+        len: usize,
```

Comment:
Why does the caller need to pass in `len` instead of using `self.len`?

Reply:
The original reason was because the calls to `ArrayVec::len` didn't get optimized at all so spliting `insert` into multiple functions actually introduced a performance loss. So, to avoid those calls, I made `len` a parameter so that the length could just be "calculated" once and then stored into a variable that gets passed around. There also was a debug assertion to make sure that `len == self.len()` but I removed it when I found useful that a vector could be treated as if it only had `len - 1` elements inside it (see usage in `force_insert`).

Reply:
That's understandable. Also, now it's not a `pub` function we can have greater confidence that `len` will be correct.

What were you using to quantify the performance loss?

Reply:
I was using the `time` command on this program:

```rust
#![feature(test)]

use const_arrayvec::ArrayVec;

fn main() {
    for _ in 0..1024 {
	test_insert();
    }
}

fn test_insert() {
    let mut v: ArrayVec<String, { 1024 }> = ArrayVec::new();

    std::hint::black_box(&mut v);

    for i in 0..1024 {
        v.insert(0, i.to_string());
    }

    std::hint::black_box(v);
}
```

And I remember noticing a `30ms` difference which immediately disappeared each time I added the `len` parameter.

I didn't remember the exact numbers so I redid the tests with `hyperfine` and there is no difference anymore (?). I'm glad we can remove the `len` parameter in `insert_unchecked` but it's weird that a difference in speed that was that consistent disappeared.

```
# With len
Benchmark #1: ./b
  Time (mean ± σ):     223.3 ms ±   1.6 ms    [User: 222.4 ms, System: 0.8 ms]
  Range (min … max):   219.8 ms … 226.2 ms    20 runs

# Without len
Benchmark #1: target/release/a
  Time (mean ± σ):     221.9 ms ±   1.4 ms    [User: 221.4 ms, System: 0.2 ms]
  Range (min … max):   219.0 ms … 225.0 ms    20 runs
```